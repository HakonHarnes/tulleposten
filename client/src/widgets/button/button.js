// @flow
import * as React from 'react';
import { Component } from 'react-simplified';

import style from './css/button.module.css';

//button used for forms 
class ButtonSubmit extends Component<{ className: string, children?: React.Node }> {
   render() {
      return (
         <button data-testid="button-submit" type="submit" className={`${this.props.className} ${style.button} ${style['button-submit']}`}>
            {this.props.children}
         </button>
      );
   }
}

//neutral button used for navigation
class ButtonNeutral extends Component<{ onClick: () => mixed, className?: string, children?: React.Node }> {
   alerts: { id: number, text: React.Node, type: string }[] = [];

   render() {
      return (
         <button
            data-testid="button-neutral"
            type="button"
            className={`${this.props.className ? this.props.className : ''} ${style.button} ${style['button-neutral']}`}
            onClick={this.props.onClick}
         >
            {this.props.children}
         </button>
      );
   }
}

//danger button used for deleting 
class ButtonDanger extends Component<{ onClick: () => mixed, className: string, children?: React.Node }> {
   render() {
      return (
         <button
            data-testid="button-danger"
            type="button"
            className={`${this.props.className} ${style.button} ${style['button-danger']}`}
            onClick={this.props.onClick}
         >
            {this.props.children}
         </button>
      );
   }
}

//menu button used in header 
class ButtonMenu extends Component<{ onClick: () => mixed, className: string, children?: React.Node }> {
   render() {
      return (
         <button
            data-testid="button-menu"
            type="button"
            alt=""
            className={`${this.props.className} ${style.button} ${style['button-menu']}`}
            onClick={this.props.onClick}
         >
            {this.props.children}
         </button>
      );
   }
}

export class Button {
   static Neutral = ButtonNeutral;
   static Submit = ButtonSubmit;
   static Danger = ButtonDanger;
   static Menu = ButtonMenu;
}
