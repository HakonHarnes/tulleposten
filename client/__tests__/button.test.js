//@flow
import * as React from 'react';
import ReactDOM from 'react-dom';

import '@testing-library/jest-dom/extend-expect'

import { render, fireEvent, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer'; 

import { Button } from '../src/widgets/button/button';

const buttonHandler = jest.fn(); 

afterEach(cleanup); 

it('renders without crashing', () => {
   const div = document.createElement('div'); 
   ReactDOM.render(<Button.Neutral className='button' onClick={() => buttonHandler()}>Click me!</Button.Neutral>, div); 
   ReactDOM.unmountComponentAtNode(div);
}); 

it('renders correctly', () => {
   const { getByTestId } = render(<Button.Neutral className='button' onClick={() => buttonHandler()}>Click me!</Button.Neutral>); 
   expect(getByTestId('button-neutral')).toHaveTextContent('Click me!'); 
   expect(getByTestId('button-neutral')).toHaveClass('button'); 
}); 

it('calls onClick function when clicked', () => {
   const { getByTestId } = render(<Button.Neutral className='button' onClick={() => buttonHandler()}>Click me!</Button.Neutral>); 

   expect(buttonHandler).not.toHaveBeenCalled(); 
   fireEvent.click(getByTestId('button-neutral')); 
   expect(buttonHandler).toHaveBeenCalled(); 
});

it('matches snapshot', () => {
   const tree = renderer.create(<Button.Neutral className='button' onClick={() => buttonHandler()}>Click me!</Button.Neutral>).toJSON(); 
   expect(tree).toMatchSnapshot(); 
}); 