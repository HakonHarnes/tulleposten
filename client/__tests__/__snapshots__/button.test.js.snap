// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  className="button button button-neutral"
  data-testid="button-neutral"
  onClick={[Function]}
  type="button"
>
  Click me!
</button>
`;
